components/script/dom/readablestream.rs

```diff
@@ -89,16 +112,18 @@ impl ReadableStream {
         obj: *mut JSObject,
         realm: InRealm,
     ) -> Result<DomRoot<ReadableStream>, ()> {
-        if !IsReadableStream(obj) {
-            return Err(());
-        }
+        /*if !IsReadableStream(obj) {
+                return Err(());
+        }*/
+
+        root_from_object(obj, *cx)
```

Comment:
I'm not sure I fully grasp the behaviors resulting from this change.

As far as I understand, before this change, there was the SpiderMonkey native `ReadableStream`s, and then there was the Servo `ReadableStream` "interface", which was only a Rust-level wrapper over native JS streams, and was never exposed to JS – and this method made a Rust `DomRoot<ReadableStream>` from the JS object. But this PR exposes objects of the Rust-level interface to JS, while `globalThis.ReadableStream` is still the SpiderMonkey implementation. So `obj` could be either a SpiderMonkey or a Servo `ReadableStream`.

The documentation for `root_from_object` says it returns `Err` if `obj` is not the right type of object. But in my tests it seems like neither `new Response(new ReadableStream())` or `new Response(new File([], "").stream())` throw. So I'm not sure what's going on. How are the fields of the Rust `ReadableStream` struct populated when it's being created from a SpiderMonkey stream?

Reply:
Great question! With these changes, we shouldn't be creating any JS stream objects; they will only be created from the native ReadableStream::new_from_bytes constructor. The only caller of ReadableStream::from_js is in generated code (https://github.com/servo/servo/blob/ffccc5c88b8d47efc777b0bc9d2620bed4bc84d7/components/script/dom/bindings/codegen/CodegenRust.py#L923), and that should only be called if we're (for example) passing a readable stream object into a DOM method that takes a union (eg. the Request/Response constructor).

Reply:
Sorry, I'm not actually sure that I really answered what you were asking. I'm actually starting to doubt myself.

Reply:
Before these changes, we actually had our Servo interface for ReadableStream that was exposed all the time, and passing `new ReadableStream()` to a function like Response's constructor called `ReadableStream::from_js` which would create a new Servo object and store the JS stream in a field. However, the actual generated code for the conversion looks like this:
```
        if value.get().is_object() {
            match ReadableStreamOrXMLHttpRequestBodyInit::TryConvertToReadableStream(SafeJSContext::from_ptr(cx), value) {
                Err(_) => return Err(()),
                Ok(Some(value)) => return Ok(ConversionResult::Success(ReadableStreamOrXMLHttpRequestBodyInit::ReadableStream(value))),
                Ok(None) => (),
            }
            match ReadableStreamOrXMLHttpRequestBodyInit::TryConvertToBlob(SafeJSContext::from_ptr(cx), value) {
                Err(_) => return Err(()),
                Ok(Some(value)) => return Ok(ConversionResult::Success(ReadableStreamOrXMLHttpRequestBodyInit::Blob(value))),
                Ok(None) => (),
            }
```
And `TryConvertToReadableStream` looks like this:
```
            match ReadableStream::from_js(cx, value.get().to_object(), InRealm::Already(&in_realm_proof)) {
                Ok(val) => val,
                Err(()) => {
                return Ok(None);
                }
            }
```
So passing a JS readable stream object into the function just means it is silently initialized with an empty body instead.

Reply:
That doesn't seem to be what's going on. `new Response(new ReadableStream()).text()` should never resolve, since the passed stream never closes, and that is the behavior you see in Servo before this PR. With this PR, however, it resolves immediately with the string `"[object ReadableStream]"`.

Since `new ReadableStream()` creates a SpiderMonkey stream, which is not an instance of Servo's `_ReadableStream` WebIDL interface, `ReadableStream::from_js` on such an object fails. JS streams are, however, convertible to strings, and that is the variant of the `BodyInit` that it goes with.

But since the whole point of this PR is to get JS streams closer to being gone, maybe living with this bug is fine until everything is moved off of JS streams.